### PR TITLE
[WIP] Add Support for masking output using a Context-Free-Grammar

### LINKED ISSUE #25778: Support for context-free-grammars (CFG) to constrain model output
### Feature request

It would be nice to constrain the model output with a CFG directly when calling `model.generate`.

This is already done by llama.cpp [grammars](https://github.com/ggerganov/llama.cpp#constrained-output-with-grammars)

An example would be defining in the following [repo](https://github.com/r2d4/rellm). 

```
prompt = "ReLLM, the best way to get structured data out of LLMs, is an acronym for "
pattern = regex.compile(r'Re[a-z]+ L[a-z]+ L[a-z]+ M[a-z]+')
output = complete_re(model=model, 
                     prompt=prompt,
                     pattern=pattern)
```
```
> Realized Logistic Logistics Model
```

Is that on the roadmap? 

### Motivation

This can be super useful to make model output parseable within architectures that process the output of an LLM using classical methods. E.g. it can be used to make a model generate valid JSON in every case. 

### Your contribution

Happy to build this with CFGs if it helps! 😄

### PR DESCRIPTION
**Pull request by jvhoffbauer**
# What does this PR do?
Fixes #25778 

# Review 
@gante @LysandreJik as discussed. Also, @oobabooga @ArthurZucker @jorgemcgomes feel free to look at this. 
This is in large parts inspired by [rellm](https://github.com/r2d4/rellm/tree/main) and [parserllm](https://github.com/r2d4/parserllm) by @r2d4 so I am tagging him here too for visibility.

# Discussion 
This is totally WIP still. I added a notebook showcasing how it might work when using the generation API from a low-level perspective. Please educate me if that is ok that way. Long-term I want to add tests. 
Generally, we use Lark to parse the grammar. The grammar itself would most likely come from a text file or a string but Lark also has a grammar object that one can use. I built a class [CfgParsingStepper](https://github.com/jvhoffbauer/transformers/blob/cfg_masking_logits_processor/src/transformers/generation/logits_process.py#L1721) that we can use to get the state of the parser for any input string. It will give us the terminal symbol we are processing and the regex for this symbol that we can use to find valid tokens. The [CfgLogitsProcessor](https://github.com/jvhoffbauer/transformers/blob/cfg_masking_logits_processor/src/transformers/generation/logits_process.py#L1774) gets the state every turn for every beam. We can consider persisting the state during generation which might be faster, instead of recalculating it. 

The whole codebase is still very much WIP. Most importantly, we still need to 

- Add tests (instead of a Jupyter Notebook) 
- Add error handling to make sure the parser throws an error if it is started with an invalid input 
- Connect the logits processor to the actual generation APIs. I am thinking of something like `model.generate(..., grammar=Grammar(...))`
- Think about cases where you want the grammar constraint to start only for the new generations and not for the input prompt

Anyhow, I wanted to put this out there for discussion. Let me know if it makes sense this way and whether you would like me to continue working in this way or if I should change anything!

### CODE REVIEW

**Comment by LysandreJik** [comment]
Thanks for your PR! I'm prompting @gante for review once he's back from leave :)

I appreciate your effort and patience @jvhoffbauer!

**Comment by jvhoffbauer** [comment]
> Thanks for your PR! I'm prompting @gante for review once he's back from leave :)
> 
> I appreciate your effort and patience @jvhoffbauer!

Hey @LysandreJik, did you already have time to review?

**Comment by oobabooga** [comment]
My 2 cents: it would be desirable to have compatibility with the BNF grammars in the llama.cpp repository:

https://github.com/ggerganov/llama.cpp/tree/master/grammars

Existing work in this direction (using custom `LogitsProcessor` in the transformers library):

https://github.com/Shopify/torch-grammar

https://github.com/im-not-tom/text-generation-webui-output-template-extension/

The original llama.cpp PR:

https://github.com/ggerganov/llama.cpp/pull/1773

**Comment by LysandreJik** [comment]
Hey @jvhoffbauer, I will let @gante that just came back from holiday review as soon as he has a spare cycle; he's the owner of `generate` and should be able to provide a much better review than I can :)

Thanks for your contribution

**Comment by gante** [comment]
(it's not forgotten, it's in my queue -- I should be able to review it over the new few days)

**Comment by jvhoffbauer** [comment]
Awesome! It’s just a draft. Please let me know your thoughts and I will focus on wrapping it up to a complete PR. Potentially already over the coming week.

**Review by gante — COMMENTED** [review]
Thank you for opening the PR @jvhoffbauer! And apologies for my delayed review 🤗 

This is a very challenging PR to review. On one hand, see the value and the possibilities it unlocks. On the other hand, it is hard to incorporate inside `generate` -- first allow me to explain the problem, to then propose an alternative avenue!

Problem: this operation depends on the `tokenizer`. We really want to avoid having tokenizer-related operations inside `generate`, as a) it will introduce a new degree of freedom and complexity to an already complex function; b) tokenization happens in the CPU, which means this will be an (unexpected) cause of slowdowns to most GPU users AND will difficult/restrict our hardware optimization goals.

Proposed alternative: In technical terms, we can work around the problem I described above if we iteratively call `generate` one token at a time, passing around `past_key_values` for inference speed purposes, and performing this additional logic outside the `generate` call. My suggestion would be to add an advanced tutorial to our `transformers` documentation, using `lark` and an example as in your notebook (FYI we intend to do the same with RAG). In other words, this wouldn't be a logits processor, but the code to add this feature would be just as easy to find as it and our team wouldn't suffer from the complexity expansion 🤗 

WDYT?

**Comment by gante** [comment]
@jvhoffbauer absolutely! You can start by drafting a stand-alone `.py` file (easier to see the diff and iterate than a notebook :) ), and when we're happy with it, I will share further pointers on integrating it into the documentation.

After this is done, I would love to invite you to write a community blog post explaining the virtues of context-free-grammar and to share a space with this technique! 💛

**Comment by jvhoffbauer** [comment]
Sounds great! I will prepare a draft and we can iterate.

**Comment by Saibo-creator** [comment]
Is it a bit overkill to introduce the dependency of Lark to implement the grammar-constrained decoding feature ? 
It seems the llama-cpp repo simply made a standalone implementation ?  @gante  https://github.com/ggerganov/llama.cpp/tree/master/grammars

**Comment by gante** [comment]
@Saibo-creator It's okay, since `lark` is stable and has no dependencies of its own :)

**Comment by Saibo-creator** [comment]
For this feature, I think there are several aspects to take into account while implementing:
1. how does this work with different tokenizers(bpe, unigram, wordpiece etc). For example, LLAMA-CPP's implementation is specific to llama tokenizer, I don't think it will work for wordpiece. 
2. unicode support. This is a bit tricky but important for multilingual usage, llama-cpp 's first implementation didn't take care of that and later they spotted and fixed it . https://github.com/ggerganov/llama.cpp/issues/2501
3.  possible comptatbility with other sampling methods such as top-k sampling

**Comment by Saibo-creator** [comment]
> My 2 cents: it would be desirable to have compatibility with the BNF grammars in the llama.cpp repository:
> 
> https://github.com/ggerganov/llama.cpp/tree/master/grammars
> 
> Existing work in this direction (using custom `LogitsProcessor` in the transformers library):
> 
> https://github.com/Shopify/torch-grammar
> 
> https://github.com/im-not-tom/text-generation-webui-output-template-extension/
> 
> The original llama.cpp PR:
> 
> [ggerganov/llama.cpp#1773](https://github.com/ggerganov/llama.cpp/pull/1773)

See  https://github.com/huggingface/transformers/pull/27557 for an implementation compatible with llama-cpp

**Comment by RadixSeven** [comment]
Is anyone still working on this? I need this functionality to test an idea. I can either do a quick fork for my own use or spend a little longer and make something worth sharing.
